hilti/src/compiler/visitors/validator.cc

```diff
@@ -139,27 +139,15 @@ struct Visitor : public visitor::PostOrder<void, Visitor> {
         auto t = n.elementType();
 
         if ( ! n.value().empty() && t == type::unknown )
-            error("non-empty list cannot have unknown type", p);
+            error("list elements have inconsistent types", p);
     }
 
     void operator()(const ctor::Map& n, position_t p) {
         auto kt = n.keyType();
         auto et = n.elementType();
 
         if ( ! n.value().empty() && (kt == type::unknown || et == type::unknown) )
-            error("non-empty map cannot have unknown type", p);
-
-        for ( const auto& [k, v] : n.value() ) {
-            if ( k.type() != kt ) {
-                error("type mismatch in map keys", p);
-                break;
-            }
-
-            if ( v.type() != et ) {
-                error("type mismatch in map values", p);
-                break;
-            }
-        }
+            error("map elements have inconsistent types", p);
     }
```

Comment:
All this does now for `Map` and also `Set` is to check for unresolved types. Do you think it would make sense to introduce a dedicated validator for that and to remove these cases here? -- if one were to remove e.g., this function one would currently hit a hard error in codegen which seems errorprone,
```
$ btest hilti.types.map.ctor-fail -d
[  0%] hilti.types.map.ctor-fail ... failed
  % 'btest-diff output' failed unexpectedly (exit code 1)
  % cat .diag
  == File ===============================
  [internal-error] <singleton>: codegen: type <unknown type> does not have a visitor

  --- Aborting
  # 0   libhilti.dylib                      0x000000010de18101 _ZN5hilti2rt9BacktraceC2Ev + 81
  # 1   libhilti.dylib                      0x000000010de18765 _ZN5hilti2rt9BacktraceC1Ev + 21
  # 2   libhilti.dylib                      0x000000010df9cdbe _ZN4util20abort_with_backtraceEv + 62
  # 3   libhilti.dylib                      0x000000010df8d116 _ZN5hilti6Logger13internalErrorERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEERKNS_8LocationE + 118
  # 4   libhilti.dylib                      0x000000010e1820bd _ZN5hilti6Logger13internalErrorINS_4type6detail4TypeELPv0EEEvNSt3__112basic_stringIcNS6_11char_traitsIcEENS6_9allocatorIcEEEERKT_ + 125
  # 5   libhilti.dylib                      0x000000010e180e94 _ZN5hilti6detail7CodeGen7compileERKNS_4type6detail4TypeENS0_7codegen9TypeUsageE + 292
  # 6   libhilti.dylib                      0x000000010e0a0d86 _ZN12_GLOBAL__N_17VisitorclERKN5hilti4ctor3MapE + 246
  # 7   libhilti.dylib                      0x000000010e09ca8e _ZN5hilti6detail7visitor15do_dispatch_oneINSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEENS_4ctor3MapENSA_6detail4CtorEN12_GLOBAL__N_17VisitorENS1_8IteratorINS_4NodeELNS1_5Or
derE0ELb0EEEEENS3_11conditionalIXsr3stdE9is_void_vIT_EEbNS3_8optionalISL_EEE4typeERT1_RKSt9type_infoRT2_RNT3_8PositionERb + 206
  ..
```

Reply:
I  don't quite follow what you mean here. It checks for unresolved types, but that includes types inside the containers just being inconsistent, which is the same the previous check did explicitly. Removing the function would of course skip this validation and then run into trouble later.

Reply:
I was hoping that the _unknown type_ property would propagate up into the container `Ctor` so that we could check that, but it seems I misunderstood since the following cannot replace all the checks of the individual type `Ctor`s, 
```
void operator()(const Ctor& c, position_t p) {
    if ( c.type() == type::unknown )
        error("cannot construct value of unspecified type", p);
}
```

Reply:
In some cases it's actually ok to have a type::unknown coming from a ctor, in particular in case of list because they may be used to intialize other types where the elements can be coerced appropriately, like in `vector<real> x = [3.14, 4];`. The list itself doesn't have a consistent type, but all elements coerce into floats, so it should be fine.

Reply:
I think I understand what you are saying, but constructs like `local auto x = vector<real>(1, 1.1)` should now be rejected.

Reply:
Well, turns out this patch missed vectors anyways, so I'm adding that. With that in place, the above wasn't rejected, but also didn't work as expected: the element coercion didn't happen. I'm working on a new version of the patch that gets that in place.